### PR TITLE
api: add runCucumber function internally

### PR DESCRIPTION
**Pull request by davidjgoss**
# Description

This adds the async `runCucumber` function for doing a test run programatically. It's not yet exposed publicly, but is used internally by the CCK tests.

This has involved:

- Pulling most of the guts out of the `Cli` class, since that is now a consumer of this new function
- Handling the i18n-related commands purely in the `cli` module
- Various other tweaks and refactors

# Motivation & context

First step towards #1711.

Now seems a good time to validate the approach on a few fronts:

- Defaulting of configuration options is generally moving _inwards_, so it's not tied to the argv parsing and so that API consumers can call with a very small signature if appropriate.
- The shape of the config object passed in to `runCucumber`. The idea is that the `support` block could be replaced with an existing support code library on subsequent runs in the same process, which solves a problem for a few framework authors.
- No breaking changes in the existing `Cli`, `Runtime` etc - framework authors are using these in the wild and we need to have a stable and compelling alternative before pulling them.

If we're happy and this goes in, the rest can follow quite quickly. Next steps:

- Implement the ability to reuse support code library
- Add a mechanism to load configuration from argv, files etc in a sane way
- Add integration tests specifically for this API
- Expose publicly (probably with an "experimental" qualifier), add documentation

## Type of change

- Refactoring/debt (improvement to code design or tooling without changing behaviour)
- New feature (non-breaking change which adds new behaviour)

# Checklist:

- [x] I have read the [**CONTRIBUTING**](../CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

### CODE REVIEW

**Review by aurelien-reeves — COMMENTED** [review]
After a first look, it looks really great! Well done!

I'll have a deeper look later

**Comment by nicojs** [comment]
Greet to see progress here! 😍

> The shape of the config object passed in to `runCucumber`. The idea is that the `support` block could be replaced with an existing support code library on subsequent runs in the same process, which solves a problem for a few framework authors.

Good to hear it is still on the radar. It will dramatically improve mutation testing performance ⏩💨

File: src/run/runCucumber.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import { IRunConfiguration } from '../configuration'
+import { IRunResult } from './types'
+import { resolvePaths } from './paths'
+import { makeRuntime } from './runtime'
+import { initializeFormatters } from './formatters'
+import { getSupportCodeLibrary } from './support'
+
+export async function runCucumber(
+  configuration: IRunConfiguration,
+  environment: {
```
Comment (mattwynne) [inline src/run/runCucumber.ts:21]: Can we name this type?

**Comment by mattwynne** [comment]
I can only offer nit-picks, this looks great to me!

**Review by aurelien-reeves — APPROVED** [review]
It looks really great!
I agree with Matt's comment.
Beside that, this is an approval 🚀

**Comment by davidjgoss** [comment]
Thanks for the feedback @aurelien-reeves @nicojs @mattwynne - I'm going to address Matt's very fair nit-pick, beef up the tests a bit and then get this merged and on to the next part.

**Comment by nicojs** [comment]
Q: does this break the current (undocumented) API? 

This is what we're using in our cucumber plugin right now:

```ts
const argv = [
      'node',
      'cucumber-js',
      '--retry',
      '0',
      '--parallel',
      '0',
      '--format',
      require.resolve('./stryker-formatter'),
      ...bailArgs,
      ...tagsArgs,
      ...profileArgs,
      ...testFilterArgs,
    ];
    const cli = new Cli({
      argv,
      cwd: process.cwd(),
      stdout: process.stdout,
    });
   await cli.run();
```

**Comment by aurelien-reeves** [comment]
> Question: does this break the current (undocumented) API?
> 
> This is what we're using in our cucumber plugin right now:
> 
> ```ts
> const argv = [
>       'node',
>       'cucumber-js',
>       '--retry',
>       '0',
>       '--parallel',
>       '0',
>       '--format',
>       require.resolve('./stryker-formatter'),
>       ...bailArgs,
>       ...tagsArgs,
>       ...profileArgs,
>       ...testFilterArgs,
>     ];
>     const cli = new Cli({
>       argv,
>       cwd: process.cwd(),
>       stdout: process.stdout,
>     });
>    await cli.run();
> ```

I took a look at the code in the branch. As far as I can tell, no, your snippet won't be broken. It will use the new API under the hood, but it will work as expected :)

**Comment by davidjgoss** [comment]
@nicojs @aurelien-reeves yep, very much a conscious choice to avoid any breaking changes to existing Cli and Runtime for the near to mid term at least.

**Comment by davidjgoss** [comment]
@binarymist I don't think `getConfiguration` was intended to be public, it perhaps should have been a private method. I think I understand your use case though having looked through the code and this should be catered for via the new API in the next RC.

**Comment by binarymist** [comment]
> this should be catered for via the new API in the next RC

Great!

Problme is... If we didn't use `getConfiguration`, then we wouldn't be able to show the _Build User_ (think end user) the test plan that they are about to run.

**Pull request merged**